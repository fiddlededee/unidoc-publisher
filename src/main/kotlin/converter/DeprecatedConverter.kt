package converter

@Deprecated("Moved to fodt package")
typealias FodtConverter = converter.fodt.FodtConverter
