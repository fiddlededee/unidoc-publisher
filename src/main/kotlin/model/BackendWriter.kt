package model
interface BackendWriter {
    // pattern visitor
    fun write(openBlock: OpenBlock)
    fun write(image: Image)
    fun write(table: Table)
    fun write(tableRowGroup: TableRowGroup)
    fun write(tr: TableRow)
    fun write(td: TableCell)
    fun write(colGroup: ColGroup)
    fun write(col: Col)
    fun write(ol: OrderedList)
    fun write(ul: UnorderedList)
    fun write(li: ListItem)
    fun write(h: Heading)
    fun write(p: Paragraph)
    fun write(doc: Document)
    fun write(text: Text)
    fun write(span: Span)
    fun write(a: Anchor)
    fun write(toc: Toc)
    fun write(textFrame: TextFrame)
    fun write(footnote: Footnote)
    fun write(footnoteRef: FootnoteRef)
    fun write(dummyNode: Node)
}
