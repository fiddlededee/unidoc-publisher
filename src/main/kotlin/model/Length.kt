package model

class Length(var value: Float, var unit: LengthUnit = LengthUnit.perc) {
    companion object Factory {
        fun fromString(length: String?): Length? {
            if (length == null) return null
            return LengthUnit.entries
                .map {
                    val inValueUnit = when (it) {
                        LengthUnit.inch -> "in"; LengthUnit.perc -> "%"
                        LengthUnit.parrots -> ""; else -> it.toString()
                    }
                    """^([0-9]+[.]?[0-9]*)(${inValueUnit})$""".toRegex()
                        .matchEntire(length) to it
                }.filter { it.first != null }
                .getOrNull(0)
                ?.run { Length(first!!.groupValues[1].toFloat(), second) }
        }
    }
}

enum class LengthUnit { cm, mm, inch, px, pt, pc, em, ex, ch, rem, vw, vh, vmin, vmax, perc, parrots, mmm }
