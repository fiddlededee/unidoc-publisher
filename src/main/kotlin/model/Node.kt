package model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator

fun document(init: Document.() -> Unit = {}): Document {
    return Document().apply(init)
}

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
abstract class Node() {

    var id: String? = null
    var sourceTagName: String? = null
    var includeTags: MutableSet<String> = mutableSetOf()
    var sourceAttributes: MutableMap<String, String> = mutableMapOf()
    var sourceMapping: SourceMapping? = null

    @get:JsonIgnore
    abstract val isInline: Boolean

    @JsonIgnore
    private var parent: Node? = null

    private val children: ArrayList<Node> = arrayListOf()
    val roles = arrayListOf<String>()
    fun <T : Node> appendChild(childNode: T): T {
        val oldParent = childNode.parent
        children.add(childNode)
        childNode.parent = this
        oldParent?.removeChild(childNode)
        return childNode
    }

    fun <T : Node> appendFirstChild(childNode: T): T {
        if (children().size == 0)
            appendChild(childNode) else
            children()[0].insertBefore(childNode)
        return childNode
    }

    fun table(init: Table.() -> Unit = {}): Table {
        return appendChild(Table().apply(init))
    }

    fun img(src: String, init: Image.() -> Unit = {}): Image {
        return appendChild(Image(src).apply(init))
    }

    fun tableRowGroup(type: TRG, init: TableRowGroup.() -> Unit = {}): TableRowGroup {
        return appendChild(TableRowGroup(type).apply(init))
    }

    fun col(width: Length): Col {
        return appendChild(Col(width))
    }

    fun tr(init: TableRow.() -> Unit = {}): TableRow {
        return appendChild(TableRow().apply(init))
    }

    fun td(init: TableCell.() -> Unit = {}): TableCell {
        return appendChild(TableCell().apply(init))
    }

    fun h(level: Int, init: Heading.() -> Unit = {}): Heading {
        return appendChild(Heading(level).apply(init))
    }


    fun li(init: ListItem.() -> Unit = {}): ListItem {
        return appendChild(ListItem().apply(init))
    }

    fun ol(init: OrderedList.() -> Unit = {}): OrderedList {
        return appendChild(OrderedList().apply(init))
    }

    fun ul(init: UnorderedList.() -> Unit = {}): UnorderedList {
        return appendChild(UnorderedList().apply(init))
    }

    fun p(init: Paragraph.() -> Unit = {}): Paragraph {
        return appendChild(Paragraph().apply(init))
    }

    fun span(init: Span.() -> Unit = {}): Span {
        return appendChild(Span().apply(init))
    }

    fun a(href: String, init: Anchor.() -> Unit = {}): Anchor {
        return appendChild(Anchor(href).apply(init))
    }

    fun text(text: String): Text {
        return appendChild(Text(text))
    }

    fun openBlock(init: OpenBlock.() -> Unit = {}): OpenBlock {
        return appendChild(OpenBlock().apply(init))
    }

    operator fun String.unaryPlus(): Text {
        return this@Node.appendChild(Text(this))
    }

    private fun removeChild(childNode: Node) {
        children.remove(childNode)
    }

    fun remove() {
        val parentNode = parent
        if (parentNode != null) {
            parentNode.removeChild(this)
        } else throw Exception("Node doesn't hava a parent")
    }

    fun insertAfter(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node after another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index + 1, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }

    fun insertBefore(nodeToInsert: Node): Node {
        val oldParent = nodeToInsert.parent
        val parent = this.parent
            ?: throw Exception("Can't add node before another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index, nodeToInsert)
        nodeToInsert.parent = parent
        oldParent?.removeChild(nodeToInsert)
        return nodeToInsert
    }


    // todo abstract ?
    abstract fun write(bw: BackendWriter)

    private fun ArrayList<Node>.fillWithDescendants(node: Node, filter: (Node) -> Boolean) {
        node.children.forEach {
            if (filter(it)) {
                add(it)
            }
            fillWithDescendants(it, filter)
        }
    }

    fun descendant(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>().apply { fillWithDescendants(this@Node, filter) }
    }

    fun ancestor(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val ancestors = arrayListOf<Node>()
        var parent = this@Node.parent()
        while (parent != null) {
            if (filter.invoke(parent)) ancestors.add(parent)
            parent = parent.parent()
        }
        return ancestors
    }

    fun nextSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val nextSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return nextSibling
        if (parent.children.indexOf(this) == parent.children.size - 1) return nextSibling
        IntRange(parent.children.indexOf(this) + 1, parent.children.size - 1).forEach {
            if (filter(parent.children[it])) nextSibling.add(parent.children[it])
        }
        return nextSibling
    }

    fun previousSibling(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        val previousSibling = arrayListOf<Node>()
        val parent = this.parent() ?: return previousSibling
        if (parent.children.indexOf(this) == 0) return previousSibling
        IntRange(0, parent.children.indexOf(this) - 1).forEach {
            if (filter(parent.children[it])) previousSibling.add(parent.children[it])
        }
        return previousSibling
    }
    fun hasNext(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != parent.children.size - 1)
    }

    fun next(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == parent.children.size - 1)
            throw Exception("The ${this::class.java.simpleName} is last")
        return parent.children[parent.children.indexOf(this) + 1]
    }

    fun hasPrevious(): Boolean {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        return (parent.children.indexOf(this) != 0)
    }

    fun previous(): Node {
        val parent = this.parent() ?: throw Exception("The ${this::class.java.simpleName} has no parent")
        if (parent.children.indexOf(this) == 0)
            throw Exception("The ${this::class.java.simpleName} is first")
        return parent.children[parent.children.indexOf(this) - 1]
    }

    fun index(): Int {
        val parent = this.parent()
            ?: throw Exception("The ${this::class.java.simpleName} node has no parent")
        return parent.children.indexOf(this)
    }

    @JsonProperty("children")
    fun getChildren(): ArrayList<Node> {
        return children
    }

    fun children(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>()
            .apply { children.forEach { if (filter(it)) this.add(it) } }
    }

    fun parent(): Node? {
        return parent
    }

    fun roles(vararg role: String) {
        role.forEach { roles.add(it) }
    }

    fun toYamlString(): String {
        return ObjectMapper(YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
            .writeValueAsString(this)
    }

    fun <T : Node> T.wrapNodeInlineContents(): T {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].insertBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> appendChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children.forEach { nodeChild ->
            if (nodeChild.isInline) itemsToWrap.last().add(nodeChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }


    // Don't understand html rendering rules, hope I'm right about this
    fun normalizeWhitespaces() {
        val localAst = this
        // Cleaning whitespaces after and before block nodes
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            val localParent = blankTextNode.parent() ?: return@forEach
            val previousConsideredBlock =
                (!blankTextNode.hasPrevious() && !localParent.isInline) ||
                        (blankTextNode.hasPrevious() && (
                                !blankTextNode.previous().isInline ||
                                        blankTextNode.previous().sourceTagName == "br"
                                ))
            val nextConsideredBlock =
                (!blankTextNode.hasNext() && !localParent.isInline) ||
                        (blankTextNode.hasNext() && (
                                !blankTextNode.next().isInline ||
                                        blankTextNode.next().sourceTagName == "br"
                                ))

            if (previousConsideredBlock || nextConsideredBlock)
                blankTextNode.remove()
        }
        // Removing extra whitespaces
        Document().apply { appendChild(localAst) }.descendant { !it.isInline }.forEach { block ->
            if (block.sourceTagName == "pre" || block.ancestor { it.sourceTagName == "pre" }.isNotEmpty())
                return@forEach
            val nodesToNormalise = block.descendant { node ->
                (node is Image || node is Text || !node.isInline) &&
                        node.ancestor { !it.isInline }.firstOrNull() == block
            }
            var chain: ArrayList<Node> = arrayListOf()
            nodesToNormalise.forEach searchForChainsToNormalize@{ node ->
                if (!node.isInline) {
                    if (chain.isEmpty()) return@searchForChainsToNormalize else {
                        normalizeChain(chain)
                        chain = arrayListOf()
                    }
                } else chain.add(node)
            }
            normalizeChain(chain)
        }
    }

    private fun normalizeChain(chain: ArrayList<Node>) {
        if (chain.isEmpty()) return
        var trimStart = true
        chain.forEach byNodesToNormalize@{ node ->
            if (node is Image || !node.isInline) trimStart = false
            if (node !is Text) return@byNodesToNormalize
            if (trimStart) {
                node.text = node.text.trimStart()
            } else {
                if (node.text.trimStart() != node.text) node.text = " " + node.text.trimStart()
            }
            if (node.text.trimEnd() != node.text && node.text.isNotEmpty()) {
                node.text = node.text.trimEnd() + " "
                trimStart = true
            } else if (node.text.isNotEmpty()) trimStart = false
        }
        val iterator = chain.reversed().iterator()
        while (iterator.hasNext()) {
            val el = iterator.next()
            if (el !is Text) break
            if (el.text.isBlank()) el.remove() else {
                el.text = el.text.trimEnd()
                break
            }
        }

    }

    fun extractText(replaceRule: (Text) -> String = { it.text }): String {
        if (this is Text) return replaceRule.invoke(this)
        return descendant { it is Text }
            .map { it as Text }
            .joinToString("") { text -> replaceRule.invoke(text) }
    }

}

