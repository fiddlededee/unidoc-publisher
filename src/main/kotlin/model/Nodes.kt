package model

import java.util.*

class DummyNode() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

abstract class NoWriterNode() : Node() {
    override fun write(bw: BackendWriter) {
        throw Exception("Writer in NoWriterNodes should be defined in CustomStyle")
    }
}

class Document() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

}

class OpenBlock() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class RectDimensions(val width: Length, val height: Length)

enum class AnchorType { ASCHAR }
class Image(
    var src: String,
    var width: Length? = null,
    var height: Length? = null,
    var anchorType: AnchorType = AnchorType.ASCHAR,
) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Col(var width: Length) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}


class ColGroup : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}


class Table() : Node() {
    // TODO: implement cols property
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

enum class TRG { head, body, foot }

class TableRowGroup(val type: TRG) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableRow() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableCell(var colspan: Int = 1, var rowspan: Int = 1) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

//    fun wrapNodeInlineContents(): TableCell {
//        // Wrapping each inline nodes chains into paragraph node
//        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
//            itemsToWrap.forEach { itemsChain ->
//                if (itemsChain.isEmpty()) return@forEach
//                itemsChain[0].addBefore(Paragraph())
//                    .apply { itemsChain.forEach { itemToWrap -> addChild(itemToWrap) } }
//            }
//        }
//
//        // Finding inline nodes chains
//        val itemsToWrap = arrayListOf<ArrayList<Node>>()
//        itemsToWrap.add(arrayListOf())
//        this.children.forEach { nodeChild ->
//            if (nodeChild.isInline) itemsToWrap.last().add(nodeChild)
//            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
//        }
//        wrap(itemsToWrap)
//        return this
//    }
}

class OrderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class UnorderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class ListItem() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

    fun wrapListItemContents(): ListItem {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].insertBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> appendChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children().forEach { listItemChild ->
            if (listItemChild.isInline) itemsToWrap.last().add(listItemChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }
}

class Header(var level: Int) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Paragraph() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

// Look like excessive
//class Pre() : Node() {
//    override val isInline: Boolean get() = false
//    override fun write(bw: BackendWriter) {
//        bw.write(this)
//    }
//}
//
//class Code() : Node() {
//    override val isInline: Boolean get() = false
//    override fun write(bw: BackendWriter) {
//        bw.write(this)
//    }
//}


class Span() : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Anchor(var href: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Text(var text: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Toc(var levels: Int = 3, title: String = "Table of contents 111") : Node() {
    var titleNode = Paragraph().apply { +title }
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TextFrame() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Footnote(
    val footnoteId: String = UUID.randomUUID().toString()
) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
    init {
        id = footnoteId
    }
}

class FootnoteRef(val footnote: Footnote) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}
