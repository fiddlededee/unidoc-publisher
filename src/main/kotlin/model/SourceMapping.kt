package model

class SourceMapping (var Lineno : Int, var file : String?)