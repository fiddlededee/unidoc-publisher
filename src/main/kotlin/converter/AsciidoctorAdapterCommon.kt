package converter

import model.*
import reader.HtmlNode
import reader.UnknownTagProcessing

object AsciidoctorAdapterCommon {
    val unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = {
        if (classNames().intersect(
                setOf(
                    "listingblock",
                    "admonitionblock",
                    "exampleblock",
                    "imageblock",
                    "paragraph",
                    "book",
                    "article",
                    "title",
                    "footnotes",
                    "footnote",
                    "dlist",
                    "olist",
                    "ulist",
                    "content",
                    *(1..6).map { "sect$it" }.toTypedArray(),
                )
            ).isNotEmpty()
        ) {
            UnknownTagProcessing.PROCESS
        } else if (setOf("dl", "dt", "dd").contains(nodeName())) {
            UnknownTagProcessing.PROCESS
        } else if (nodeName() == "div") {
            UnknownTagProcessing.PASS
        } else UnknownTagProcessing.UNDEFINDED
    }

    @JvmStatic
    fun Node.wrapTitleInlineContent() {
        this.descendant { it is OpenBlock && it.roles.contains("title") }
            .forEach { openBlock ->
                openBlock.wrapNodeInlineContents()
                val table = openBlock.parent()
                if (table != null && table is Table) {
                    val tableId = table.id
                    if (tableId != null) {
                        openBlock.descendant { it is Paragraph }.firstOrNull()?.apply {
                            id = tableId
                        }
                        table.id = "table--${tableId}"
                    }
                    table.insertBefore(openBlock)
                    openBlock.roles("table-title")
                }
            }
    }

    fun Node.wrapBlockImages() {
        this.descendant { image ->
            image is Image &&
                    image.ancestor { it.roles.contains("imageblock") }
                        .isNotEmpty()
        }.forEach { image ->
            val imageBlock = image.ancestor { it.roles.contains("imageblock") }.firstOrNull()
            val p =
                Paragraph().apply { roles("imageblock-paragraph") }
            if (imageBlock?.id != null) {
                p.id = imageBlock.id
                imageBlock.id = "image-block--${imageBlock.id}"
            }
            image.insertAfter(p)
            p.appendChild(image)
        }
    }


    @JvmStatic
    fun Node.extractToc() {
        val parsedToc = this.descendant { it.id == "toc" }
            .firstOrNull() ?: return
        val toc = Toc().apply {
            levels = parsedToc.descendant { it is UnorderedList }
                .maxOfOrNull { ul -> ul.ancestor { it is UnorderedList }.size + 1 }
                ?: throw Exception("Can't determine number of levels in the TOC")
            titleNode = Paragraph().apply {
                parsedToc.descendant { it.parent()?.id == "toctitle" }.forEach {
                    appendChild(it)
                }
            }
        }
        parsedToc.insertBefore(toc)
        parsedToc.remove()
    }

    fun Node.transformAdmonitions() {
        this.descendant { it.roles.contains("admonitionblock") }.forEach { block ->
            val oldTable = block.children().first { it is Table }
            val (typeContent, content) = run {
                oldTable
                    .children().first { it is TableRowGroup }
                    .children().first { it is TableRow }
                    .apply {
                        if (children().size != 2 ||
                            children().any { it !is TableCell }
                        ) throw Exception("Can't define admonition block")
                    }.children()
            }
            typeContent.roles("admonition-type-content")
            content.roles("admonition-content")
            block.apply {
                table {
                    col(Length(100F))
                    tableRowGroup(TRG.body) {
                        tr { appendChild(typeContent) }
                        tr { appendChild(content) }
                    }
                }
            }
            oldTable.remove()
        }
    }


}