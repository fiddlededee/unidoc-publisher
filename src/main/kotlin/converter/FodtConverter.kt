package converter

import common.GenericAdapter
import fodt.FodtGenerator
import model.Document
import model.Node
import model.Text
import org.redundent.kotlin.xml.PrintOptions
import reader.GenericHtmlReader
import reader.HtmlNode
import reader.UnknownTagProcessing
import writer.OdWriter
import writer.OdtStyleList

open class FodtConverter(init: FodtConverter.() -> Unit) {
    private var adapter: GenericAdapter? = null
    fun adaptWith(adapter: GenericAdapter) {
        unknownTagProcessingRule = adapter.unknownTagProcessingRule
        odtStyleList = adapter.basicStyle()
        this.adapter = adapter
    }

    var html: String? = null
        set(value) {
            val fieldNotNullable = value ?: throw Exception("html value can't be null")
            field = fieldNotNullable; htmlNode = HtmlNode(fieldNotNullable)
        }

    var xpath: String = "/html/body"
    var unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing = { UnknownTagProcessing.UNDEFINDED }
    var customNodeProcessingRule: Node.(htmlNode: HtmlNode) -> Unit = {}
    var odtStyleList = OdtStyleList()
    var template: String? = null
    var ast: Node? = null
    var preList: ArrayList<PreRegion> = arrayListOf()
    var fodtGenerator: FodtGenerator? = null
    var fodt: String? = null
    private var htmlNode: HtmlNode? = null

    init {
        this.apply(init)
    }

    fun ast(): Node {
        return ast ?: throw Exception("AST is empty")
    }

    fun template(): String {
        return template ?: throw Exception("Template is empty")
    }

    fun preList(): ArrayList<PreRegion> {
        if (preList.isEmpty()) throw Exception("The list of pre FODT elements is empty")
        return preList
    }

    fun html(): String {
        return html ?: throw Exception("HTML is emplty")
    }

    fun fodt(): String {
        return fodt ?: throw Exception("Fodt is empty")
    }

    fun parse() {
        val localHtmlNode = htmlNode?.selectAtXpath(xpath)
            ?: throw Exception("Please set html")
        ast = Document()
        val localAst = ast ?: throw Exception("Error: ast variable was mutated")
        newReaderInstance(
            localAst, localHtmlNode,
            unknownTagProcessingRule, customNodeProcessingRule
        )
            .apply { localAst.setBasics(localHtmlNode); iterateAll() }
        // Don't understand html rendering rules, hope I'm right about this
        localAst.descendant { it is Text && it.text.isBlank() }.forEach { blankTextNode ->
            if (!blankTextNode.hasPrevious() ||
                !blankTextNode.previous().isInline ||
                blankTextNode.previous().roles.contains("br") ||
                !blankTextNode.hasNext() ||
                !blankTextNode.next().isInline ||
                blankTextNode.next().roles.contains("br")
            )
                blankTextNode.remove()
        }
    }

    fun generatePre() {
        val localAst = ast ?: throw Exception("Ast is null")
        adapter?.apply { localAst.normalizeAll() }
        (localAst.descendant { it.includeTags.isNotEmpty() } +
                localAst
                ).forEach { nodeToProcess ->
                val includeTags = nodeToProcess.includeTags
                    .let { if (it.isNotEmpty()) it.toSet() else setOf("all") }
                val pre = newOdWriterInstance(odtStyleList = odtStyleList)
                    .apply { (nodeToProcess.write(this)) }
                    .preOdNode.toString(PrintOptions(pretty = false))
                preList.add(PreRegion(includeTags, pre, nodeToProcess.isInline))
            }

    }

    fun generateFodt() {
        val localTemplate = template ?: throw Exception("Please, set template variable")
        if (preList.isEmpty()) throw Exception("Please set pre variable")
        fodtGenerator = FodtGenerator(preList, localTemplate)
        serializeFodt()
    }

    fun convert() {
        parse(); generatePre(); generateFodt()
    }

    fun ast2fodt() {
        generatePre(); generateFodt()
    }

    fun serializeFodt() {
        fodt = fodtGenerator?.serialize()
    }

    open fun newReaderInstance(
        ast: Node, htmlNode: HtmlNode,
        unknownTagProcessingRule: HtmlNode.() -> UnknownTagProcessing,
        customNodeProcessing: Node.(htmlNode: HtmlNode) -> Unit
    ): GenericHtmlReader {
        return GenericHtmlReader(ast, htmlNode, unknownTagProcessingRule, customNodeProcessing)
    }

    open fun newOdWriterInstance(odtStyleList: OdtStyleList): OdWriter {
        return OdWriter(odtStyleList = odtStyleList)
    }
}