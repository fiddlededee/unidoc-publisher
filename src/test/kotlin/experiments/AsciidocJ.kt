package experiments

import model.Text
import org.approvaltests.Approvals
import org.asciidoctor.Asciidoctor
import org.asciidoctor.Attributes
import org.asciidoctor.Options
import org.asciidoctor.ast.ContentNode
import org.asciidoctor.ast.Document
import org.asciidoctor.ast.Section
import org.asciidoctor.ast.StructuralNode
import org.asciidoctor.converter.ConverterFor
import org.asciidoctor.converter.StringConverter
import org.asciidoctor.jruby.ast.impl.PhraseNodeImpl
import org.asciidoctor.log.LogRecord
import org.asciidoctor.log.Severity
import org.junit.jupiter.api.Test
import java.util.UUID


class AsciidocJ {

    @Test
    fun useConverterDirectly() {
        val asciidocText = """
            == Section *One*
            
            Some *text*!
            
        """.trimIndent()

        val factory: Asciidoctor = Asciidoctor.Factory.create()
        factory.javaConverterRegistry().register(TextConverter::class.java)
        val modelDocument = model.Document()
        val attributes = Attributes.builder()
            .attribute("unidoc-publisher-model", modelDocument)
            .build()
        factory.convert(
            asciidocText, Options.builder()
                .backend("text")
                .attributes(attributes)
                .build()
        )
        // Placing inline nodes to appropriate places
        modelDocument.descendant { it.id?.startsWith("inline-nodes-") ?: false }
            .forEach { inlineNode ->
                val inlineId = inlineNode.id!!.substringAfter("inline-nodes-")
                modelDocument.descendant { it is Text && it.text.contains(inlineId) }
                    .map { it as Text }
                    .firstOrNull()
                    ?.apply {
                        val beforeAfter = text.split("@@${inlineId}@@")
                        arrayOf(Text(beforeAfter[0]), inlineNode, Text(beforeAfter[1]))
                            .filter { it !is Text || it.text.isNotBlank() }
                            .forEach { insertBefore(it) }
                        inlineNode.id = null
                        remove()
                    }
            }
        Approvals.verify(modelDocument.toYamlString())
    }
}


@ConverterFor("text")
class TextConverter(backend: String?, opts: Map<String?, Any?>?) :
    StringConverter(backend, opts) {
    private var modelSet = false
    private lateinit var unidocModel: model.Document

    override fun convert(
        node: ContentNode, transform: String?, o: Map<Any, Any>
    ): String? {
        if (!modelSet) {
            unidocModel = node.document.attributes["unidoc-publisher-model"] as model.Document
            modelSet = true
        }
        var localTransform: String? = transform
        if (localTransform == null) {
            localTransform = node.nodeName
        }

        if (node is Document) {
            node.content
            return "dummy"
        } else if (node is Section) {
            unidocModel.apply {
                h(node.level) {
                    +node.title
                }
                node.content
            }
            return "dummy"
        } else if (localTransform == "paragraph") {
            val block = node as StructuralNode
            val content = block.content as String
            unidocModel.apply { p { +content } }
            return "dummy"
        } else if (node is PhraseNodeImpl) {
            val uid = UUID.randomUUID()
            unidocModel.apply {
                if (node.type == "strong")
                    span { roles("strong"); id = "inline-nodes-$uid"; +node.text }
            }
            return ("@@${uid}@@")
        } else {
            log(
                LogRecord(
                    Severity.WARN,
                    "Unexpected node ${node::class.java.simpleName}, transform (${transform})"
                )
            )
        }
        return null
    }
}
