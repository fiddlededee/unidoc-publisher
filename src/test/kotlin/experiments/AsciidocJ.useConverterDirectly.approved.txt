!<Document>
children:
- !<Header>
  level: 1
  children:
  - !<Text>
    text: "Section "
  - !<Span>
    roles:
    - "strong"
    children:
    - !<Text>
      text: "One"
- !<Paragraph>
  children:
  - !<Text>
    text: "Some "
  - !<Span>
    roles:
    - "strong"
    children:
    - !<Text>
      text: "text"
  - !<Text>
    text: "!"
